content/en/blog/basics-for-passwordless/index.md

```diff
@@ -0,0 +1,136 @@
+
+## To Buy or To Build?
+
+Choosing one of many authentication service providers may be the most straightforward path on your
+passwordless journey. A service provider typically gives you access to a backend as
+a SaaS or a product you host yourself. The backend has the above capabilities to store and verify
+your users' public keys. In addition to backend functionality, the service providers offer
+client libraries that enable you to add matching authenticator support to your application,
+whether a web or a native application.
+
+{{< imgproc platform Fit "825x825" >}}
+<em>Platform and browser handle the user interaction dialogs when using authenticators.</em>
+{{< /imgproc >}}
+
+Another option is to implement the WebAuthn support yourself. Some excellent [open-source libraries](https://github.com/herrjemand/awesome-webauthn)
+already exist that ease the development of your backend service's functionality for
+public key handling. [Browser support for WebAuthn capabilities](https://caniuse.com/webauthn) is
+rather good, and the integration to the web application is straightforward
+once the backend is in place. One can utilize dedicated
+client libraries for native applications (for example, [iOS](https://developer.apple.com/documentation/authenticationservices/public-private_key_authentication)
+and [Android](https://developer.android.com/training/sign-in/passkeys)).
+
+## Example of Getting Hands Dirty
+
+In our OSS decentralized identity agency project, we implemented WebAuthn logic ourselves.
+We created the authentication functionality in our backend server and web application.
+The core functionality consists of two features: registering a new user and authenticating
+an existing user (and returning an access token after a successful authentication). Of course,
+an end-product would have more features, e.g., for adding and removing authenticators
+but starting with the core features is the simplest.
+
+We wrote the authentication server in Go.
+It utilizes [`go-webauthn`](https://github.com/go-webauthn/webauthn) library.
+Both of the core features need two API endpoints.
+On the client side, we use
+the [navigator credentials JavaScript API](https://w3c.github.io/webappsec-credential-management/#framework-credential-management)
+in a React application. The following sequence graphs demonstrate in more detail how the logic flows:
+
+**Registration**
+
+```mermaid
+sequenceDiagram
+    autonumber
+    participant Client
+    participant Server
+    Client->>Server: Request for credential creation options.<br/>/attestation/options
+    Server-->>Client: Return credential creation options.
+    Note left of Client: Create credential with received options.<br/>navigator.credentials.create
+    Client->>Server: Send client data and public key to server.<br/>/attestation/result
+    Note right of Server: Validate data. <br/> Store the public key and the credential id, </br> and attach to user.
+    Server-->>Client: Registration OK!
+```
+
```

Comment:
I suggest adding some text here. It doesn't seem very easy to read the sequence diagrams when they are together. The first time, I thought they were one large scenario.

Reply:
Yes, I am doubting if those diagrams are needed at all, but would like to describe somehow in more details, what exactly needs to be done if one wishes to implement this from scratch 🤔